Main menu

IgnacioAbadie opened the pull request:
## Summary ##

https://trello.com/c/FECsllw2/11-n-41-table-ui-mock-data


Added Navigation action from login button to Main Menu
Added Mocked table view for books list

## Screenshots ##

### iPhone se
<img width="882" alt="Captura de Pantalla 2020-01-21 a la(s) 15 23 49" src="https://user-images.githubusercontent.com/24418053/72831819-5e03f280-3c62-11ea-8720-d978092b19c1.png">
<img width="504" alt="Captura de Pantalla 2020-01-21 a la(s) 15 23 45" src="https://user-images.githubusercontent.com/24418053/72831820-5e03f280-3c62-11ea-9904-08435b175fca.png">



### iPhone 11 pro
<img width="1048" alt="Captura de Pantalla 2020-01-21 a la(s) 15 22 47" src="https://user-images.githubusercontent.com/24418053/72831834-63f9d380-3c62-11ea-9571-a0517691f0c3.png">
<img width="584" alt="Captura de Pantalla 2020-01-21 a la(s) 15 22 51" src="https://user-images.githubusercontent.com/24418053/72831835-64926a00-3c62-11ea-8911-26c4822894ff.png">

IgnacioAbadie committed:
Added Table view for books

IgnacioAbadie committed:
Added spacing between rows

IgnacioAbadie committed:
fixed xib styles by code to add spacing between rows

IgnacioAbadie committed:
Added Navigation from login to Main

gmazzei commented on WBooks/Screens/CustomBookCell.swift:
```diff
@@ -0,0 +1,36 @@
+        self.backgroundColor = UIColor.clear
+    }
+
+    override func setSelected(_ selected: Bool, animated: Bool) {
```
Remove this method if it only calls `super`.

gmazzei commented on WBooks/Screens/CustomBookCell.swift:
```diff
@@ -0,0 +1,36 @@
+    
+    override func awakeFromNib() {
+        super.awakeFromNib()
+        self.backgroundColor = UIColor.clear
```
Omit the `self`'s that are not necessary. You can omit `UIColor.` as well:
```Swift
backgroundColor = .clear
```

IgnacioAbadie replied:
oh! nice

gmazzei commented on WBooks/Screens/MainViewController.swift:
```diff
@@ -0,0 +1,60 @@
+class MainViewController: UIViewController, UITableViewDelegate, UITableViewDataSource {
+    private let _view: MainView = MainView.loadFromNib()!
+    
+    var booksArray : Array = [["image": "img_book2", "title": "The best book in the world", "subtitle": "Peter Sjermstrom"], ["image": "img_book3", "title": "A little bird told me", "subtitle": "Timpthy Cross"], ["image": "img_book4", "title": "When the doves desappeared", "subtitle": "Sofi Oksanen"]]
```
Make this `private let`.

gmazzei commented on WBooks/Screens/MainViewController.swift:
```diff
@@ -0,0 +1,60 @@
+    }
+    
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: "CustomBookCell", for: indexPath) as! CustomBookCell
```
Avoid using forced unwrapping whenever possible. You can use `guard` instead.
```Swift
guard let cell = tableView.dequeueReusableCell(withIdentifier: "CustomBookCell", for: indexPath) as? CustomBookCell else { return UITableViewCell() }
```

gmazzei commented on WBooks/Screens/MainViewController.swift:
```diff
@@ -0,0 +1,60 @@
+        super.viewDidLoad()
+        self._view.booksTable.delegate = self
+        self._view.booksTable.dataSource = self
+        let nib = UINib.init(nibName: "CustomBookCell", bundle: nil)
```
Omit the `init` word. You can do `UINib(nibName: "CustomBookCell", bundle: nil)` instead. 
Also remove unnecessary `self`'s.

gmazzei commented on WBooks/Screens/MainViewController.swift:
```diff
@@ -0,0 +1,60 @@
+        self._view.booksTable.delegate = self
+        self._view.booksTable.dataSource = self
+        let nib = UINib.init(nibName: "CustomBookCell", bundle: nil)
+        self._view.booksTable.register(nib, forCellReuseIdentifier: "CustomBookCell")
```
Since `"CustomBookCell"` is used multiple times, it would be better to move it to a static constant, like so:
```Swift
private static let cellReuseIdentifier = "CustomBookCell"
```

gmazzei commented on WBooks/Screens/MainViewController.swift:
```diff
@@ -0,0 +1,60 @@
+        
+        let rowData = booksArray[indexPath.row]
+        
+        cell.imageSource.image = UIImage(named: rowData["image"]!) 
```
Use safe unwrapping here too. You can use `if let` instead.

```Swift
if let imageName = rowData["image"] {
    cell.imageSource.image = UIImage(named: imageName)
}
```

gmazzei commented on WBooks/Screens/LoginViewController.swift:
```diff
@@ -30,5 +30,12 @@ class LoginViewController: UIViewController {
+    @objc private func buttonTapped() {
+        let mainVC = MainViewController()
+        mainVC.modalPresentationStyle = .fullScreen
+        present(mainVC, animated: false, completion: nil)
```
Let's call this `mainViewController` instead of `mainVC`.

gmazzei commented on WBooks/Screens/MainViewController.swift:
```diff
@@ -0,0 +1,60 @@
+import UIKit
+import WolmoCore
+
+class MainViewController: UIViewController, UITableViewDelegate, UITableViewDataSource {
```
Some good practices when extending multiple protocols are:

1) Using "MARK:" as a heading for different sections

```Swift
class MainViewController: UIViewController, UITableViewDelegate, UITableViewDataSource {    

   // MARK: - UITableViewDelegate

   [Here you add methods related with UITableViewDelegate...]

   // MARK: - UITableViewDataSource
      
   [Here you add methods related with UITableViewDataSource...]
}
```

2) You can also split the class into different sections using extensions:

```Swift
class MainViewController: UIViewController {
    // Code related to UIViewController
}

extension MainViewController: UITableViewDelegate {
    // Here you add methods related with UITableViewDelegate...
}

extension MainViewController: UITableViewDataSource {
    // Here you add methods related with UITableViewDataSource...
}
```

IgnacioAbadie replied:
Nice! i'll try

gmazzei replied:
Actually, the "MARK:" for UIViewController and MainViewController are not necessary, because "MARK:" is mainly used for protocols. Anyway, everything looks great! 🙂

carcos-wolox commented on WBooks/Screens/MainViewController.swift:
```diff
@@ -0,0 +1,60 @@
+    }
+    
+    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
+        100
```
Is this size enough if for example, the name of the book takes 3 lines?

IgnacioAbadie replied:
Title is set for two lines max

![image](https://user-images.githubusercontent.com/24418053/72839156-8266cb80-3c70-11ea-9af9-f8b05d131a9a.png)


subtitle just one line

Thats ok?

carcos-wolox replied:
You can make the cell grow based on it's content, but it's ok for this

carcos-wolox commented on WBooks/Screens/MainViewController.swift:
```diff
@@ -0,0 +1,60 @@
+import UIKit
+import WolmoCore
+
+class MainViewController: UIViewController, UITableViewDelegate, UITableViewDataSource {
```
Try to keep inits at the beginning of the class

IgnacioAbadie committed:
Fixed code style

- Removed unnnecesary self.
- refactor MainViewController to separate extensions functions.
- added safe and guard unwrapped.

carcos-wolox commented on WBooks/Screens/MainViewController.swift:
```diff
@@ -0,0 +1,60 @@
+    }
+    
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: "CustomBookCell", for: indexPath) as! CustomBookCell
```
We have some helpful functions in wolmo core for this. 
Check register and dequeue funcs [here](https://github.com/Wolox/wolmo-core-ios/blob/master/WolmoCore/Extensions/UIKit/UITableView.swift)

IgnacioAbadie replied:
nice!

carcos-wolox commented on WBooks/Screens/MainViewController.swift:
```diff
@@ -0,0 +1,60 @@
+        
+        cell.imageSource.image = UIImage(named: rowData["image"]!) 
+        cell.titleLabel.text = rowData["title"]
+        cell.subtitleLabel.text = rowData["subtitle"]
```
You could have a func in `CustomBookCel` for this configuration

IgnacioAbadie replied:
Nice, i like it

IgnacioAbadie committed:
extracted configuration cell to appropiated class and used WolmoCore utils for cells

IgnacioAbadie committed:
used wolmo core register for uiTableView and extended custom cell from Nibloadable

gmazzei reviewed (approved):
Nice work! 🏆

gmazzei commented on WBooks/Screens/CustomBookCell.swift:
```diff
@@ -0,0 +1,38 @@
+    
+    override func awakeFromNib() {
+        super.awakeFromNib()
+        backgroundColor = .clear
```
I think you can set `selectionStyle = .none` to prevent the cell from changing its color when tapped.

carcos-wolox commented on WBooks/Screens/CustomBookCell.swift:
```diff
@@ -0,0 +1,38 @@
+        backgroundColor = .clear
+    }
+    
+    func setBookAttributes(data: [String : String]) {
```
Maybe with a Model would be better (You are going to do that later) ... if you want, change it now 😄

IgnacioAbadie replied:
TODO 🤣

carcos-wolox commented on WBooks/Screens/MainViewController.swift:
```diff
@@ -0,0 +1,66 @@
+    private static let cellReuseIdentifier = "CustomBookCell"
+    
+    private let _view: MainView = MainView.loadFromNib()!
+    
```
Unnecessary line

carcos-wolox commented on WBooks/Screens/MainViewController.swift:
```diff
@@ -0,0 +1,66 @@
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        booksArray.count
+    }
+    
```
extra line

carcos-wolox reviewed (approved):
Just some style-comments. But LGTM 🎉

IgnacioAbadie committed:
removed extra empty lines and added selectionStyle to none
